pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ThresholdLoadManagerStrategy.java

```diff
@@ -0,0 +1,133 @@
+            return Double.POSITIVE_INFINITY;
+        }
+
+        double brokerResourceUsage = brokerData.getLocalData().getMaxResourceUsageWithWeight(
+                conf.getLoadBalancerCPUResourceWeight(), conf.getLoadBalancerMemoryResourceWeight(),
+                conf.getLoadBalancerDirectMemoryResourceWeight(), conf.getLoadBalancerBandwithInResourceWeight(),
+                conf.getLoadBalancerBandwithOutResourceWeight());
+
+        if (log.isDebugEnabled()) {
+            log.debug("Broker {} resource usage {}",
+                    brokerData.getLocalData().getWebServiceUrl(), brokerResourceUsage);
+        }
+        return brokerResourceUsage;
+    }
+
+    /**
+     * Find a suitable broker to assign the given bundle to.
+     *
+     * @param candidates
+     *            The candidates for which the bundle may be assigned.
+     * @param bundleToAssign
+     *            The data for the bundle to assign.
+     * @param loadData
+     *            The load data from the leader broker.
+     * @param conf
+     *            The service configuration.
+     * @return The name of the selected broker as it appears on ZooKeeper.
+     */
+    @Override
+    public Optional<String> selectBroker(final Set<String> candidates, final BundleData bundleToAssign,
+                                         final LoadData loadData,
+                                         final ServiceConfiguration conf) {
+        bestBrokers.clear();
+        brokerUsages.clear();
+
+        // Maintain of list of all the brokers below average load, and then randomly
+        // select one of them at the end.
+        double sumScore = 0.0;
+        for (String broker : candidates) {
+            final BrokerData brokerData = loadData.getBrokerData().get(broker);
+            final double score = getScore(brokerData, conf);
+            if (score == Double.POSITIVE_INFINITY) {
+                final LocalBrokerData localData = brokerData.getLocalData();
+                log.warn(
+                        "Broker {} is overloaded: CPU: {}%, MEMORY: {}%, DIRECT MEMORY: {}%, BANDWIDTH IN: {}%, "
+                                + "BANDWIDTH OUT: {}%",
+                        broker, localData.getCpu().percentUsage(), localData.getMemory().percentUsage(),
+                        localData.getDirectMemory().percentUsage(), localData.getBandwidthIn().percentUsage(),
+                        localData.getBandwidthOut().percentUsage());
+                continue;
+            }
+            sumScore += score;
+            brokerUsages.put(broker, score);
+        }
+
+        if (brokerUsages.isEmpty()) {
+            // All brokers are overloaded.
+            // Assign randomly in this case.
+            bestBrokers.addAll(candidates);
+        } else {
+            double avgUsage = sumScore / brokerUsages.size();
+            for (Map.Entry<String, Double> entry : brokerUsages.entrySet()) {
+                if (entry.getValue() <= avgUsage) {
```

Comment:
I wonder what's benefit of random select from brokers below average than select the least loaded broker?

Reply:
Always selecting the least loaded broker can cause the chosen broker to quickly become heavily loaded, especially when there is a lot of unloaded traffic in a short time. Random select can avoid this.

Reply:
@dragonls it makes sense to do random selection. It seems that this is an important detail which is missing from the PR description. Please add this information to the description of the algorithm.

Reply:
Updated.